chapters/05-comparison.Rmd

```diff
@@ -27,170 +27,174 @@ distributions_to_evaluate = [
+
+```{python compareSimpleBinnedFFTISJRuntimeInstantiation, echo=FALSE, eval=TRUE, fig.cap="Runtime difference of the instantiaton phase between the five algorithms 'Exact', 'Binned', 'FFT', 'ISJ' and 'Hofmeyr'"}
+figure, axes = runner.plot_runtimes(restricted_runtimes, distributions_to_evaluate, ['ZfitExact', 'ZfitBinned', 'ZfitFFT', 'ZfitISJ', 'ZfitHofmeyrK1withCpp'], 'instantiation')
 figure
 ```
 
-Although the ISJ and the FFT based approach are slower to instantiate, they are significantly faster for larger datasets during the PDF evaluation step.
+As seen in figure \@ref(fig:compareSimpleBinnedFFTISJRuntimeInstantiation), the FFT and ISJ method use more time during the instantiation phase than the other methods. This is expected, since for these methods the kernel density estimate is calculated for every grid point during the instantiation phase, whereas for the other methods, the calculation is only prepared and actually executed during the evaluation phase itself. In addition, we can see that the FFT method is faster than the ISJ method in calculating the kernel density estimate for the grid points. The linear binning method is slower than the exact method because the bin counts are calculated  during the instantiation phase.
 
-```{python compareSimpleBinnedFFTISJRuntimePDF, echo=FALSE, eval=TRUE, fig.cap="Runtime difference of the evaluation step between the four algorithms 'Exact', 'Binned', 'FFT' and 'ISJ'", fig.align='center', out.width='100%'}
-figure, axes = runner.plot_runtimes(restricted_runtimes, distributions_to_evaluate, ['ZfitExact', 'ZfitBinned', 'ZfitFFT', 'ZfitISJ'], 'pdf')
+```{python compareSimpleBinnedFFTISJRuntimePDF, echo=FALSE, eval=TRUE, fig.cap="Runtime difference of the evaluation phase between the five algorithms 'Exact', 'Binned', 'FFT', 'ISJ' and 'Hofmeyr'", fig.align='center', out.width='100%'}
+figure, axes = runner.plot_runtimes(restricted_runtimes, distributions_to_evaluate, ['ZfitExact', 'ZfitBinned', 'ZfitFFT', 'ZfitISJ', 'ZfitHofmeyrK1withCpp'], 'pdf')
 figure
 ```
 
+In figure \@ref(fig:compareSimpleBinnedFFTISJRuntimePDF) we can see that evaluation runtime of the exact KDE method increases with increased bin size. Whereas for the other methods it stays nearly constant. The binned method benefits from the fact that, no matter how big the sample size is, it has to compute the kernel density estimate only for the fixed bin count of $N=1024$. The other methods are faster during the evaluation phase, because they have already calculated estimate in the instantiation phase and only need to interpolate for the values in between.
 
-## Comparison to KDEpy on CPU
+```{python compareSimpleBinnedFFTISJRuntimeTotal, echo=FALSE, eval=TRUE, fig.cap="Total runtime difference (instantiation and evaluation phase combined) between the five algorithms 'Exact', 'Binned', 'FFT', 'ISJ' and 'Hofmeyr'", fig.align='center', out.width='100%'}
```

Comment:
I would remove this. It's just the above two added, but this doesn't really add something as the instantiation is called once, the evaluation n times. Basically, the instantiation time, if same order of magnitude as evaluation time, can be forgotten about.

The plot doesn't add as we are interested in the instantiation + n times evaluation time. But what is n? We don't know, it depends on the case. We just know that n >> 1, e.g. we can often forget about the instantiation time